stew/base36.nim

```diff
@@ -0,0 +1,238 @@
+## Copyright (c) 2025 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+## This module implements Base36 encoding and decoding procedures.
+## This module supports two variants of Base36 encoding (Bitcoin and Flickr).
+
+type
+  Base36Status* {.pure.} = enum
+    Error,
+    Success,
+    Incorrect,
+    Overrun
+
+  Base36Alphabet* = object
+    decode*: array[128, int8]
+    encode*: array[36, uint8]
+
+  Base36* = object
+  Base36Uc* = object
+  Base36Lc* = object
+
+  Base36Types* = Base36 | Base36Uc | Base36Lc
+
+  Base36Error* = object of CatchableError
+    ## Base36 specific exception type
+
+func newAlphabet36(s: string): Base36Alphabet =
+  doAssert(len(s) == 36)
+  var alphabet: Base36Alphabet
+  for i in 0..<len(s):
+    alphabet.encode[i] = cast[uint8](s[i])
+  for i in 0..<len(alphabet.decode):
+    alphabet.decode[i] = -1
+  for i in 0..<len(alphabet.encode):
+    alphabet.decode[int(alphabet.encode[i])] = int8(i)
+  return alphabet
+
+const
+  B36UcAlphabet* = newAlphabet36("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ")
+  B36LcAlphabet* = newAlphabet36("0123456789abcdefghijklmnopqrstuvwxyz")
+
+func encodedLength(btype: typedesc[Base36Types], length: int): int =
+  ## Return estimated length of Base36 encoded value for plain length
+  ## ``length``.
+  return (length * 277) div 179 + 1
+
+func decodedLength(btype: typedesc[Base36Types], length: int): int =
+  ## Return estimated length of decoded value of Base36 encoded value of length
+  ## ``length``.
+  return (length * 179) div 277
+
+func encode*(btype: typedesc[Base36Types], inbytes: openArray[byte],
+             outstr: var openArray[char], outlen: var int): Base36Status =
+  ## Encode array of bytes ``inbytes`` using Base36 encoding and store
+  ## result to ``outstr``. On success ``Base36Status.Success`` will be returned
+  ## and ``outlen`` will be set to number of characters stored inside of
+  ## ``outstr``. If length of ``outstr`` is not enough then
+  ## ``Base36Status.Overrun`` will be returned and ``outlen`` will be set to
+  ## number of characters required.
+
+  when (btype is Base36) or (btype is Base36Lc):
+    const alphabet = B36LcAlphabet
+  elif (btype is Base36Uc):
+    const alphabet = B36UcAlphabet
+
+  let binsz = len(inbytes)
+  var zcount = 0
+
+  while zcount < binsz and inbytes[zcount] == 0x00'u8:
+    inc(zcount)
+
+  let size = ((binsz - zcount) * 277) div 179 + 1
+  var buffer = newSeq[uint8](size)
+
+  var hi = size - 1
+  var i = zcount
+  var j = size - 1
+  while i < binsz:
```

Comment:
isn't this just a `for i in zcount ..< binsz` loop?

It's not clear to me why desugar it to that extent, even if that's sort of what it compiles into.

It just unconditionally `inc(i)`'s exactly once each loop iteration, there's no tricky control flow where a counting `for` loop doesn't straightforwardly work because only sometimes increments its loop counter.

Reply:
https://github.com/status-im/nim-stew/compare/89498a6349addf3205cfefff8642f3d6f0f748c0..3c74cf03aea444e4ac462bf04ea957d4e6f3dd53